tests/bluetooth/bsim_bt/bsim_test_audio/src/tbs_client_test.c

```diff
@@ -0,0 +1,232 @@
+/*
+ * Copyright (c) 2019 Bose Corporation
+ * Copyright (c) 2021 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifdef CONFIG_BT_TBS_CLIENT
+
+#include <bluetooth/audio/tbs.h>
+
+#include "common.h"
+
+static struct bt_conn_cb conn_callbacks;
+extern enum bst_result_t bst_result;
+static volatile bool bt_init;
+static volatile bool is_connected;
+static volatile bool discovery_complete;
+static volatile bool is_gtbs_found;
+static volatile bool read_complete;
+static volatile bool call_placed;
+static volatile uint8_t call_state;
+static volatile uint8_t call_index;
+static struct bt_conn *tbs_conn;
+
+static void tbs_client_call_states_cb(struct bt_conn *conn, int err,
+				      uint8_t index, uint8_t call_count,
+				      const struct bt_tbs_client_call_state *call_states)
+{
+	if (index != 0) {
+		return;
+	}
+
+	printk("%s\n", __func__);
+	printk("Index %u\n", index);
+	if (err != 0) {
+		FAIL("Call could not read call states (%d)\n", err);
+		return;
+	}
+
+	call_index = call_states[0].index;
+	call_state = call_states[0].state;
+	printk("call index %u - state %u\n", call_index, call_state);
+}
+
+static void tbs_client_read_bearer_provider_name(struct bt_conn *conn, int err,
+						 uint8_t index,
+						 const char *value)
+{
+	if (err != 0) {
+		FAIL("Call could not read bearer name (%d)\n", err);
+		return;
+	}
+
+	printk("Index %u\n", index);
+	printk("Bearer name pointer: %p\n", value);
+	printk("Bearer name: %s\n", value);
+	read_complete = true;
+}
+
+static void tbs_client_discover_cb(struct bt_conn *conn, int err,
+				   uint8_t tbs_count, bool gtbs_found)
+{
+	printk("%s\n", __func__);
+	if (err != 0) {
+		FAIL("TBS_CLIENT could not be discovered (%d)\n", err);
+		return;
+	}
+
+	is_gtbs_found = true;
+	discovery_complete = true;
+}
+
+static struct bt_tbs_client_cb tbs_client_cbs = {
```

Comment:
Agreed. @asbjornsabo @Casper-Bonde-Bose @MariuszSkamra  Do you think we'll ever need to have more than 1 user registering LE Audio callbacks? In other words, should we use a list of callbacks, or a singleton?

Reply:
LE Audio callbacks in general, or TBS callbacks?
I believe a single user will be by far the most common case.  I suggest we stay with that until we see a need for anything else.

(Changing to a list will also mean significant changes elsewhere, which, if we want to do them, should be done after we have got the code int.)

Reply:
I agree. 

If we change it for TBS, it would make sense to change it across the LE Audio stack. 

regardless, I'll go ahead and add `const` to the callbacks as suggested by @jhedberg

Reply:
Changing it to `const` would require an API change here, and if we go with the `list` route `const` will need to be removed again. 

@jhedberg Can we keep as is for now?

Reply:
Created https://github.com/zephyrproject-rtos/zephyr/issues/43913 to revisit this in a more general manner

Reply:
> Changing it to const would require an API change here, and if we go with the list route const will need to be removed again.

No, I don't think we'd take that path. Currently there are (if I counted correctly) 22 callback pointers in this struct. That's a non-trivial amount of memory. So if you want to add support for multiple instances in the future you'd either do that by an additional level of indirection (a top-level struct containing the linked list pointer and then a pointer to the const callbacks struct) or with the help of a dedicated linker section placed in ROM.

I don't see what's the problem here - adding const in a couple of places in this PR is a trivial thing to do.

Reply:
> No, I don't think we'd take that path. Currently there are (if I counted correctly) 22 callback pointers in this struct. That's a non-trivial amount of memory. 

Correct, currently it is large. That can be significantly improved once https://github.com/zephyrproject-rtos/zephyr/issues/41216 is solved. 

> So if you want to add support for multiple instances in the future you'd either do that by an additional level of indirection (a top-level struct containing the linked list pointer and then a pointer to the const callbacks struct)

I don't see how that approach will provide any better memory. You'll still need a pointer for each callback function. 

> with the help of a dedicated linker section placed in ROM.

I am definitely fine with using ROM for the callbacks as well. 

Possibly we also want to move this into the `bt_tbs` struct, so that there is a callback struct for each TBS instance. 

> I don't see what's the problem here - adding const in a couple of places in this PR is a trivial thing to do.

I'll add them in this PR, and then when we revisit the callbacks we can continue the discussion there.

Reply:
> I don't see how that approach will provide any better memory. You'll still need a pointer for each callback function.

The struct with the callbacks would be placed in ROM, while the “parent” struct (with two pointers) would reside in RAM